src/_guides/language/effective-dart/design_migrated.md

```diff
@@ -1218,90 +1289,167 @@ var names = people.map((person) => person.name);
 {% prettify dart tag=pre+code %}
-Set<String> things = Set<String>();
+class Downloader {
+  final Completer<String> response = Completer<String>();
+}
 {% endprettify %}
 
-Here, the type annotation on the variable is used to infer the type argument of
-constructor call in the initializer.
-
-In other contexts, there isn't enough information to infer the type and then you
-should write the type argument:
+Here, the type annotation on the field provides a surrounding context to infer
+the type argument of constructor call in the initializer.
 
 {:.good}
 <?code-excerpt "design_good.dart (explicit)"?>
 {% prettify dart tag=pre+code %}
-var things = Set<String>();
+var items = Future.value([1, 2, 3]);
 {% endprettify %}
 
 {:.bad}
 <?code-excerpt "design_bad.dart (explicit)"?>
 {% prettify dart tag=pre+code %}
-var things = Set();
+var items = Future<List<int>>.value(<int>[1, 2, 3]);
 {% endprettify %}
 
-Here, since the variable has no type annotation, there isn't enough context to
-determine what kind of `Set` to create, so the type argument should be provided
-explicitly.
+Here, the types of the collection and instance can be inferred bottom-up from
+their elements and arguments.
 
 
-### DO annotate when Dart infers the wrong type.
+### AVOID writing incomplete generic types.
 
-Sometimes, Dart infers a type, but not the type you want. For example, you may
-want a variable's type to be a supertype of the initializer's type so that you
-can later assign some other sibling type to the variable:
+When writing a type annotation or type argument, it's still possible to not pin
+down a complete type by writing the name of a generic type but omitting its type
+arguments. In Java, these are called "raw types". For example:
 
-{:.good}
-<?code-excerpt "design_good.dart (inferred-wrong)" replace="/num highest/[!num!] highest/g"?>
+{:.bad}
+<?code-excerpt "design_bad.dart (incomplete-generic)"?>
 {% prettify dart tag=pre+code %}
-num highScore(List<num> scores) {
-  [!num!] highest = 0;
-  for (var score in scores) {
-    if (score > highest) highest = score;
-  }
-  return highest;
-}
+List numbers = [1, 2, 3];
+var completer = Completer<Map>();
 {% endprettify %}
 
-Here, if we didn't annotate a type for `highest`, then Dart would infer `int`
-from the initialized value `0`. We need the type to be `num` since we may later
-assign double values from `scores` to it. In cases like this, explicit
-annotations make sense.
+Here, `numbers` has a type annotation, but the annotation doesn't provide a type
+argument to the generic `List`. Likewise, the `Map` type argument to `Completer`
+isn't fully specified. In cases like this, Dart will *not* try to "fill in" the
+rest of the type for you using the surrounding context. Instead, it silently
+fills in any missing type arguments with with `dynamic` (or the bound if the
+class has one). That's rarely what you want.
 
+Instead, if you are writing a generic type in a type annotation or as a type
```

Comment:
This sentence is a bit hard to parse. Here's an attempt at making it a bit clearer grammatically:

```suggestion
Instead, if you're writing a generic type either in a type annotation or as a type
```